File: SAAF/java_template/src/main/java/lambda/Load.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import saaf.Inspector;
import saaf.Response;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.Statement;
import java.sql.ResultSet;
import java.text.SimpleDateFormat;
import java.util.HashMap;
import java.util.Properties;

/**
 * AWS Lambda Load Function - Service #2
 * Loads transformed CSV data from S3 into Aurora MySQL database
 * 
 * @author Team12
 */
public class Load implements RequestHandler<Request, HashMap<String, Object>> {
    
    // Database connection parameters loaded from db.properties
    private String url;
    private String username;
    private String password;
    
    /**
     * Load database configuration from db.properties file
     */
    private void loadDatabaseConfig(LambdaLogger logger) throws Exception {
        logger.log("[DEBUG] Starting to load db.properties");
        Properties prop = new Properties();
        InputStream input = Load.class.getClassLoader().getResourceAsStream("db.properties");
        
        if (input == null) {
            logger.log("[ERROR] db.properties file not found!");
            throw new Exception("Unable to find db.properties");
        }
        
        logger.log("[DEBUG] db.properties file found, loading properties");
        prop.load(input);
        
        url = prop.getProperty("url");
        username = prop.getProperty("username");
        password = prop.getProperty("password");
        
        logger.log("[DEBUG] DB Config loaded - URL: " + url + ", Username: " + username);
        
        input.close();
    }
    
    @Override
    public HashMap<String, Object> handleRequest(Request request, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("[DEBUG] ========== LOAD FUNCTION STARTED ==========");
        
        Inspector inspector = new Inspector();
        inspector.inspectAll();
        
        String bucketName = request.getBucketname();
        String fileName = request.getFilename();
        
        logger.log("[DEBUG] Request received - Bucket: " + bucketName + ", File: " + fileName);
        
        int rowsLoaded = 0;
        int duplicatesSkipped = 0;
        Connection conn = null;
        
        try {
            // Load database configuration
            logger.log("[DEBUG] Step 1: Loading database configuration");
            loadDatabaseConfig(logger);
            logger.log("[DEBUG] Step 1: COMPLETE - Database configuration loaded");
            
            // Download transformed CSV from S3
            logger.log("[DEBUG] Step 2: Starting S3 download");
            inspector.addTimeStamp("s3DownloadStart");
            AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
            logger.log("[DEBUG] S3 client created");
            
            S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketName, fileName));
            logger.log("[DEBUG] S3 object retrieved");
            
            InputStream objectData = s3Object.getObjectContent();
            BufferedReader reader = new BufferedReader(new InputStreamReader(objectData));
            inspector.addTimeStamp("s3DownloadEnd");
            logger.log("[DEBUG] Step 2: COMPLETE - S3 download finished");
            
            // Connect to Aurora MySQL
            logger.log("[DEBUG] Step 3: Attempting database connection to: " + url);
            logger.log("[DEBUG] This may take 10-30 seconds if database is asleep...");
            inspector.addTimeStamp("dbConnectionStart");
            
            conn = DriverManager.getConnection(url, username, password);
            
            logger.log("[DEBUG] Step 3: COMPLETE - Database connection established!");
            conn.setAutoCommit(false);
            logger.log("[DEBUG] Auto-commit disabled for batch processing");
            inspector.addTimeStamp("dbConnectionEnd");
            
            // Create table if it doesn't exist
            logger.log("[DEBUG] Step 4: Creating table if not exists");
            createTable(conn, logger);
            logger.log("[DEBUG] Step 4: COMPLETE - Table ready");
            
            // Prepare INSERT statement with IGNORE to skip duplicates
            logger.log("[DEBUG] Step 5: Preparing INSERT statement");
            String insertSQL = "INSERT IGNORE INTO sales_data (region, country, item_type, sales_channel, " +
                    "order_priority, order_date, order_id, ship_date, units_sold, unit_price, " +
                    "unit_cost, total_revenue, total_cost, total_profit, order_processing_time, " +
                    "gross_margin) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
            
            PreparedStatement pstmt = conn.prepareStatement(insertSQL);
            logger.log("[DEBUG] Step 5: COMPLETE - INSERT statement prepared");
            
            // Skip header line
            String line = reader.readLine();
            logger.log("[DEBUG] Header line skipped: " + line);
            
            // Load data
            logger.log("[DEBUG] Step 6: Starting data load");
            inspector.addTimeStamp("dataLoadStart");
            int batchSize = 0;
            int batchLimit = 1000;
            int totalRows = 0;
            int lineNumber = 1;
            
            while ((line = reader.readLine()) != null) {
                lineNumber++;
                
                if (line.trim().isEmpty()) {
                    continue;
                }
                
                String[] fields = parseCSVLine(line);
                
                if (fields.length < 16) {
                    logger.log("[WARNING] Skipping malformed line " + lineNumber + " (expected 16 fields, got " + 
                        fields.length + ")");
                    continue;
                }
                
                try {
                    // Set parameters for prepared statement
                    pstmt.setString(1, fields[0].trim());
                    pstmt.setString(2, fields[1].trim());
                    pstmt.setString(3, fields[2].trim());
                    pstmt.setString(4, fields[3].trim());
                    pstmt.setString(5, fields[4].trim());
                    pstmt.setString(6, convertDateFormat(fields[5].trim()));
                    pstmt.setInt(7, Integer.parseInt(fields[6].trim()));
                    pstmt.setString(8, convertDateFormat(fields[7].trim()));
                    pstmt.setInt(9, Integer.parseInt(fields[8].trim()));
                    pstmt.setDouble(10, Double.parseDouble(fields[9].trim()));
                    pstmt.setDouble(11, Double.parseDouble(fields[10].trim()));
                    pstmt.setDouble(12, Double.parseDouble(fields[11].trim()));
                    pstmt.setDouble(13, Double.parseDouble(fields[12].trim()));
                    pstmt.setDouble(14, Double.parseDouble(fields[13].trim()));
                    pstmt.setInt(15, Integer.parseInt(fields[14].trim()));
                    pstmt.setDouble(16, Double.parseDouble(fields[15].trim()));
                    
                    pstmt.addBatch();
                    batchSize++;
                    totalRows++;
                    
                    // Execute batch when limit reached
                    if (batchSize >= batchLimit) {
                        logger.log("[DEBUG] Executing batch of " + batchSize + " rows");
                        int[] results = pstmt.executeBatch();
                        conn.commit();
                        
                        for (int result : results) {
                            if (result > 0) rowsLoaded++;
                            else if (result == 0) duplicatesSkipped++;
                        }
                        
                        logger.log("[DEBUG] Batch complete. Total loaded: " + rowsLoaded + 
                            ", Duplicates: " + duplicatesSkipped);
                        batchSize = 0;
                    }
                    
                } catch (Exception e) {
                    logger.log("[ERROR] Error processing line " + lineNumber + ": " + e.getMessage());
                }
            }
            
            // Execute remaining batch
            if (batchSize > 0) {
                logger.log("[DEBUG] Executing final batch of " + batchSize + " rows");
                int[] results = pstmt.executeBatch();
                conn.commit();
                
                for (int result : results) {
                    if (result > 0) rowsLoaded++;
                    else if (result == 0) duplicatesSkipped++;
                }
                
                logger.log("[DEBUG] Final batch complete");
            }
            
            inspector.addTimeStamp("dataLoadEnd");
            logger.log("[DEBUG] Step 6: COMPLETE - Data load finished");
            
            // Cleanup
            pstmt.close();
            reader.close();
            logger.log("[DEBUG] Resources closed");
            
            // Create response
            String successMessage = "Successfully loaded data from " + bucketName + "/" + fileName + 
                ". Rows loaded: " + rowsLoaded + ", Duplicates skipped: " + duplicatesSkipped + 
                ", Total rows processed: " + totalRows;
            
            logger.log("[SUCCESS] " + successMessage);
            
            Response response = new Response();
            response.setValue(successMessage);
            
            // Add load metrics
            inspector.addAttribute("rowsLoaded", rowsLoaded);
            inspector.addAttribute("duplicatesSkipped", duplicatesSkipped);
            inspector.addAttribute("totalRowsProcessed", totalRows);
            inspector.addAttribute("bucketName", bucketName);
            inspector.addAttribute("fileName", fileName);
            inspector.addAttribute("dbUrl", url);
            
            inspector.consumeResponse(response);
            
        } catch (Exception e) {
            logger.log("[CRITICAL ERROR] Exception caught: " + e.getClass().getName());
            logger.log("[CRITICAL ERROR] Message: " + e.getMessage());
            
            // Print full stack trace
            logger.log("[STACK TRACE START]");
            for (StackTraceElement element : e.getStackTrace()) {
                logger.log("  at " + element.toString());
            }
            logger.log("[STACK TRACE END]");
            
            inspector.addAttribute("loadError", e.getMessage());
            
            // Rollback on error
            try {
                if (conn != null) {
                    logger.log("[DEBUG] Attempting rollback");
                    conn.rollback();
                    logger.log("[DEBUG] Rollback successful");
                }
            } catch (Exception rollbackEx) {
                logger.log("[ERROR] Error during rollback: " + rollbackEx.getMessage());
            }
        } finally {
            try {
                if (conn != null && !conn.isClosed()) {
                    logger.log("[DEBUG] Closing database connection");
                    conn.close();
                    logger.log("[DEBUG] Database connection closed");
                }
            } catch (Exception e) {
                logger.log("[ERROR] Error closing connection: " + e.getMessage());
            }
        }
        
        logger.log("[DEBUG] ========== LOAD FUNCTION FINISHED ==========");
        inspector.inspectAllDeltas();
        return inspector.finish();
    }
    
    /**
     * Creates the sales_data table if it doesn't exist
     */
    private void createTable(Connection conn, LambdaLogger logger) throws Exception {
        logger.log("[DEBUG] Creating sales_data table if not exists");
        
        String createTableSQL = "CREATE TABLE IF NOT EXISTS sales_data (" +
                "region VARCHAR(100), " +
                "country VARCHAR(100), " +
                "item_type VARCHAR(100), " +
                "sales_channel VARCHAR(50), " +
                "order_priority VARCHAR(50), " +
                "order_date DATE, " +
                "order_id INT PRIMARY KEY, " +
                "ship_date DATE, " +
                "units_sold INT, " +
                "unit_price DECIMAL(10,2), " +
                "unit_cost DECIMAL(10,2), " +
                "total_revenue DECIMAL(12,2), " +
                "total_cost DECIMAL(12,2), " +
                "total_profit DECIMAL(12,2), " +
                "order_processing_time INT, " +
                "gross_margin DECIMAL(5,4), " +
                "INDEX idx_region (region), " +
                "INDEX idx_country (country), " +
                "INDEX idx_item_type (item_type), " +
                "INDEX idx_order_priority (order_priority)" +
                ")";
        
        Statement stmt = conn.createStatement();
        stmt.execute(createTableSQL);
        conn.commit();
        stmt.close();
        
        logger.log("[DEBUG] Table creation/verification complete");
    }
    
    /**
     * Check if table has data and delete it if found
     * Returns the number of rows that were deleted
     */
    private int checkAndClearData(Connection conn, LambdaLogger logger) throws Exception {
        logger.log("[DEBUG] Checking row count in sales_data table");
        
        Statement stmt = conn.createStatement();
        ResultSet rs = stmt.executeQuery("SELECT COUNT(*) as row_count FROM sales_data");
        
        int rowCount = 0;
        if (rs.next()) {
            rowCount = rs.getInt("row_count");
        }
        rs.close();
        
        logger.log("[DEBUG] Found " + rowCount + " existing rows in table");
        
        if (rowCount > 0) {
            logger.log("[DEBUG] Deleting existing data from sales_data table");
            stmt.execute("DELETE FROM sales_data");
            conn.commit();
            logger.log("[DEBUG] Successfully deleted " + rowCount + " rows");
        }
        
        stmt.close();
        return rowCount;
    }
    
    /**
     * Parse CSV line handling commas within quoted fields
     */
    private String[] parseCSVLine(String line) {
        return line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1);
    }
    
    /**
     * Convert date from M/d/yyyy format to yyyy-MM-dd for MySQL
     */
    private String convertDateFormat(String dateStr) {
        try {
            SimpleDateFormat inputFormat = new SimpleDateFormat("M/d/yyyy");
            SimpleDateFormat outputFormat = new SimpleDateFormat("yyyy-MM-dd");
            return outputFormat.format(inputFormat.parse(dateStr));
        } catch (Exception e) {
            return dateStr;
        }
    }
}

File: SAAF/python_template/src/load.py
from decimal import Decimal
import boto3
import pymysql
import csv
import io
import os
import re
from datetime import datetime
from Inspector import Inspector   # your provided Inspector code

def parse_jdbc_url(jdbc_url):
    match = re.match(
        r"jdbc:mysql://([^:/]+):(\d+)/(.+)", jdbc_url
    )
    if not match:
        raise ValueError("Invalid JDBC URL format")

    return {
        "host": match.group(1),
        "port": int(match.group(2)),
        "database": match.group(3)
    }

# ---------- Database Config ----------
def load_database_config():
    jdbc_url = os.environ["DB_URL"]
    parsed = parse_jdbc_url(jdbc_url)

    try:
        db_cfg = {
            "host": parsed["host"],
            "port": parsed["port"],
            "database": parsed["database"],
            "user": os.environ["DB_USERNAME"],
            "password": os.environ["DB_PASSWORD"],
            "autocommit": False
        }
        return db_cfg
    except Exception as e:
        raise Exception(f"Database configuration error: {str(e)}")


# ---------- CSV Helpers ----------
CSV_SPLIT_REGEX = re.compile(r',(?=(?:[^"]*"[^"]*")*[^"]*$)')


def parse_csv_line(line):
    return CSV_SPLIT_REGEX.split(line)


def convert_date_format(date_str):
    try:
        return datetime.strptime(date_str, "%m/%d/%Y").strftime("%Y-%m-%d")
    except Exception:
        return date_str


# ---------- Table Creation ----------
CREATE_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS sales_data (
    region VARCHAR(100),
    country VARCHAR(100),
    item_type VARCHAR(100),
    sales_channel VARCHAR(50),
    order_priority VARCHAR(50),
    order_date DATE,
    order_id INT PRIMARY KEY,
    ship_date DATE,
    units_sold INT,
    unit_price DECIMAL(10,2),
    unit_cost DECIMAL(10,2),
    total_revenue DECIMAL(12,2),
    total_cost DECIMAL(12,2),
    total_profit DECIMAL(12,2),
    order_processing_time INT,
    gross_margin DECIMAL(5,4),
    INDEX idx_region (region),
    INDEX idx_country (country),
    INDEX idx_item_type (item_type),
    INDEX idx_order_priority (order_priority)
)
"""


# ---------- Lambda Handler ----------
def lambda_handler(event, context):
    inspector = Inspector()
    inspector.inspectAll()

    bucket_name = event["bucketname"]
    file_name = event["filename"]

    rows_loaded = 0
    duplicates_skipped = 0
    total_rows = 0
    conn = None

    try:
        # Load DB config
        db_cfg = load_database_config()

        # ---------- S3 Download ----------
        inspector.addTimeStamp("s3DownloadStart")
        s3 = boto3.client("s3")
        s3_object = s3.get_object(Bucket=bucket_name, Key=file_name)
        csv_data = s3_object["Body"].read().decode("utf-8")
        inspector.addTimeStamp("s3DownloadEnd")

        reader = io.StringIO(csv_data)
        lines = reader.readlines()

        # ---------- DB Connection ----------
        inspector.addTimeStamp("dbConnectionStart")
        conn = pymysql.connect(**db_cfg)
        cursor = conn.cursor()
        inspector.addTimeStamp("dbConnectionEnd")

        # Create table
        cursor.execute(CREATE_TABLE_SQL)
        conn.commit()

        insert_sql = """
        INSERT IGNORE INTO sales_data (
            region, country, item_type, sales_channel, order_priority,
            order_date, order_id, ship_date, units_sold, unit_price,
            unit_cost, total_revenue, total_cost, total_profit,
            order_processing_time, gross_margin
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """

        batch = []
        batch_limit = 1000

        inspector.addTimeStamp("dataLoadStart")

        # Skip header
        for line in lines[1:]:
            if not line.strip():
                continue

            fields = parse_csv_line(line.strip())
            if len(fields) < 16:
                continue

            try:
                record = (
                    fields[0].strip(),
                    fields[1].strip(),
                    fields[2].strip(),
                    fields[3].strip(),
                    fields[4].strip(),
                    convert_date_format(fields[5].strip()),
                    int(fields[6]),
                    convert_date_format(fields[7].strip()),
                    int(fields[8]),
                    float(fields[9]),
                    float(fields[10]),
                    float(fields[11]),
                    float(fields[12]),
                    float(fields[13]),
                    int(fields[14]),
                    float(fields[15]),
                )

                batch.append(record)
                total_rows += 1

                if len(batch) >= batch_limit:
                    affected = cursor.executemany(insert_sql, batch)
                    conn.commit()
                    rows_loaded += affected
                    duplicates_skipped += len(batch) - affected
                    batch.clear()

            except Exception as row_err:
                context.getLogger().log(f"Row error: {row_err}")

        # Final batch
        if batch:
            affected = cursor.executemany(insert_sql, batch)
            conn.commit()
            rows_loaded += affected
            duplicates_skipped += len(batch) - affected

        full_url = f"jdbc:mysql://{db_cfg['host']}:{db_cfg['port']}/{db_cfg['database']}"

        inspector.addTimeStamp("dataLoadEnd")

        # ---------- Metrics ----------
        inspector.addAttribute("rowsLoaded", rows_loaded)
        inspector.addAttribute("duplicatesSkipped", duplicates_skipped)
        inspector.addAttribute("totalRowsProcessed", total_rows)
        inspector.addAttribute("bucketName", bucket_name)
        inspector.addAttribute("fileName", file_name)
        inspector.addAttribute("dbUrl", full_url)
        inspector.addAttribute("loadError", Decimal("0"))

    except Exception as e:
        if conn:
            conn.rollback()
        raise

    finally:
        if conn:
            conn.close()

    inspector.inspectAllDeltas()
    return inspector.finish()